stan/math/rev/mat/functor/algebra_solver.hpp

```diff
@@ -0,0 +1,287 @@
+     * is a vector of parameters (var). The overload calls the
+     * algebraic solver defined above and builds a vari object on
+     * top, using the algebra_solver_vari class.
+     *
+     * @tparam F type of equation system function.
+     * @tparam T  Type of elements in y vectors.
+     * @param[in] f Functor that evaluates the system of equations.
+     * @param[in] x Vector of starting values.
+     * @param[in] y parameter vector for the equation system.
+     * @param[in] dat continuous data vector for the equation system.
+     * @param[in] dat_int integer data vector for the equation system.
+     * @param[in, out] msgs the print stream for warning messages.
+     * @param[in] relative_tolerance determines the convergence criteria
+     *            for the solution.
+     * @param[in] function_tolerance determines whether roots are acceptable.
+     * @param[in] max_num_steps  maximum number of function evaluations.
+     * @return theta Vector of solutions to the system of equations.
+     * @throw <code>std::invalid_argument</code> if x has size zero.
+     * @throw <code>std::invalid_argument</code> if y has size zero.
+     * @throw <code>std::invalid_argument</code> if x has non-finite elements.
+     * @throw <code>std::invalid_argument</code> if y has non-finite elements.
+     * @throw <code>std::invalid_argument</code> if dat has non-finite elements.
+     * @throw <code>std::invalid_argument</code> if dat_int has non-finite elements.
+     * @throw <code>std::invalid_argument</code> if relative_tolerance is strictly
+     * negative.
+     * @throw <code>std::invalid_argument</code> if function_tolerance is strictly
+     * negative.
+     * @throw <code>std::invalid_argument</code> if max_num_steps is not positive.
+     * @throw <code>error_type>?</code> if solver exceeds max_num_steps.
+     * @throw <code>error_type>?</code> if the norm of the solution exceeds the
+     * function tolerance.
+     */
+    template <typename F, typename T>
+    Eigen::Matrix<T, Eigen::Dynamic, 1>
+    algebra_solver(const F& f,
+                   const Eigen::VectorXd& x,
+                   const Eigen::Matrix<T, Eigen::Dynamic, 1>& y,
+                   const std::vector<double>& dat,
+                   const std::vector<int>& dat_int,
+                   std::ostream* msgs = 0,
+                   double relative_tolerance = 1e-10,
+                   double function_tolerance = 1e-6,
+                   long int max_num_steps = 1e+3) {  // NOLINT(runtime/int)
+      Eigen::VectorXd theta_dbl = algebra_solver(f, x, value_of(y), dat,
+                                                 dat_int, 0,
+                                                 relative_tolerance,
+                                                 function_tolerance,
+                                                 max_num_steps);
+
+      // FIX ME - the next three lines are redundant (they occur in the prim
```

Comment:
It's redundant in that I had an opportunity to cache the result from the call to algebra_solver, a few lines above. If both functions were in a class this would be a straightforward fix. I tried passing `fx` as an optional reference argument. It got surprisingly messy, and didn't compile. I may try again after I've cleaned up the rest of the code

Anyways, I updated the comment.